conda_lockfiles/loaders/utils.py

```diff
@@ -0,0 +1,59 @@
+from typing import Any
+
+from conda.base.context import context
+from conda.core.package_cache_data import PackageCacheData, ProgressiveFetchExtract
+from conda.exceptions import (
+    CondaExitZero,
+    DryRunExit,
+)
+from conda.models.match_spec import MatchSpec
+from conda.models.records import PackageRecord
+
+CondaPackageURL = str
+CondaPackageMetadata = dict[str, Any]
+
+
+def _matchspec_from_url_and_metadata(
+    url: CondaPackageURL, metadata: CondaPackageMetadata
+):
+    checksums = {}
+    for checksum_name in ["md5", "sha256"]:
+        if checksum_name in metadata:
+            checksums[checksum_name] = metadata[checksum_name]
+    return MatchSpec(url, **checksums)
+
+
+def records_from_urls_and_metadata(
+    metadata_by_url: dict[CondaPackageURL, CondaPackageMetadata],
+    dry_run: bool = context.dry_run,
+    download_only: bool = context.download_only,
+) -> tuple[PackageRecord, ...]:
+    fetch_specs = [
+        _matchspec_from_url_and_metadata(url, metadata)
+        for url, metadata in metadata_by_url.items()
+    ]
+    if dry_run:
+        raise DryRunExit()
+
+    pfe = ProgressiveFetchExtract(fetch_specs)
+    pfe.execute()
+
+    if download_only:
+        raise CondaExitZero(
+            "Package caches prepared. Installed cancelled with --download-only option."
+        )
+
+    records: list[PackageRecord] = []
+    for fetch_spec in fetch_specs:
+        cache_record = next(PackageCacheData.query_all(fetch_spec), None)
+        if cache_record is None:
+            raise AssertionError(f"Missing package cache record for: {fetch_spec}")
+        url = fetch_spec.get("url")
+        overrides = metadata_by_url.get(url, {})
+        records.append(
+            PackageRecord.from_objects(
+                cache_record,
+                **overrides,
+            )
+        )
+    return records
```

Comment:
`records` is currently not a tuple, also what if we just make this a generator instead?

```suggestion
    for fetch_spec in fetch_specs:
        cache_record = next(PackageCacheData.query_all(fetch_spec), None)
        if cache_record is None:
            raise AssertionError(f"Missing package cache record for: {fetch_spec}")
        url = fetch_spec.get("url")
        overrides = metadata_by_url.get(url, {})
        yield PackageRecord.from_objects(
            cache_record,
            **overrides,
        )
```

Reply:
A generator would not check that all packages are in cache before yielding a record. That check is important, if any fail (which is highly unlikely) the entire process should stop. 

Additionally, I can't come up with a good case where the caller of this this functional would not need the full set of records. Producing them one-by-one would be undone by the caller.

I'm going to cast records to a tuple in the return. This can be revisited later if needed.